### ISSUE TITLE
crash with NULL pointer in zmq::mtrie_t::match

### ISSUE THREAD

**Issue by WallStProg** [zeromq/libzmq#3166]
I've been seeing an ocassional crash in the regression tests I'm running with our application, with similar stack traces.

The root cause of the crash is the following code:

```
#0  0x0000003a6aa325e5 in raise () from /lib64/libc.so.6
#1  0x0000003a6aa33d4d in abort () from /lib64/libc.so.6
#2  0x00007f2a2e63627a in TransactX::defaultCoreSigHandler (sig=<optimized out>) at MISigHandler.cpp:528
#3  <signal handler called>
#4  zmq::mtrie_t::match (this=0x1ba9f30, data_=0x1c162e8 "_NAMING", size_=<optimized out>, func_=0x7f2a2c609c50 <zmq::xpub_t::mark_as_matching(zmq::pipe_t*, void*)>, 
    arg_=0x1ba9a70) at /build/temp/libzmq/4.2.3/src/mtrie.cpp:395
#5  0x00007f2a2c609c18 in zmq::xpub_t::xsend (this=0x1ba9a70, msg_=0x7f2a2890f400) at /build/temp/libzmq/4.2.3/src/xpub.cpp:238
#6  0x00007f2a2c5f0a34 in zmq::socket_base_t::send (this=0x1ba9a70, msg_=0x7f2a2890f400, flags_=0) at /build/temp/libzmq/4.2.3/src/socket_base.cpp:1125
#7  0x00007f2a2c610c66 in s_sendmsg (flags_=<optimized out>, msg_=<optimized out>, s_=<optimized out>) at /build/temp/libzmq/4.2.3/src/zmq.cpp:375
#8  zmq_send (s_=0x1ba9a70, buf_=0x7f2a2890f4a0, len_=<optimized out>, flags_=0) at /build/temp/libzmq/4.2.3/src/zmq.cpp:409
#9  0x00007f2a2c8468ff in zmqBridgeMamaTransportImpl_sendEndpointsMsg (impl=0x1b91b40, command=<optimized out>) at /build/temp/OpenMAMAzmq/2.0/src/transport.c:1461
#10 0x00007f2a2c849776 in zmqBridgeMamaTransportImpl_dispatchNamingMsg (impl=0x1b91b40, zmsg=<optimized out>) at /build/temp/OpenMAMAzmq/2.0/src/transport.c:558
#11 0x00007f2a2c84accb in zmqBridgeMamaTransportImpl_dispatchThread (closure=<optimized out>) at /build/temp/OpenMAMAzmq/2.0/src/transport.c:484
#12 0x0000003a6ae07aa1 in start_thread () from /lib64/libpthread.so.0
#13 0x0000003a6aae8aad in clone () from /lib64/libc.so.6

```

Running gdb on the core file reveals that `current` is NULL at the point where it access `current->pipes`: 

```
void zmq::mtrie_t::match (unsigned char *data_, size_t size_,
    void (*func_) (pipe_t *pipe_, void *arg_), void *arg_)
{
    mtrie_t *current = this;
    while (true) {

        //  Signal the pipes attached to this node.
        if (current->pipes) {
            for (pipes_t::iterator it = current->pipes->begin ();
                  it != current->pipes->end (); ++it)
                func_ (*it, arg_);
        }

```

I realize that a repro using pure ZeroMQ code would be ideal, but I suspect that will be difficult to achieve -- there's a lot of stuff going on in the library code that calls ZeroMQ that would be difficult to recreate.

So, at this point I'm just hoping for some hints from the experts here, specifically if anyone can help answer the following:

- Is there a legitimate way that `current` can be assigned a NULL value?  

I'm logging socket events using zmq_socket_monitor, so I can tell that the socket specified in the `zmq_send` call (**namingPub**) is connected, but the handshake is not complete at the time of the crash.  In fact, I get the HANDSHAKE_SUCCEEDED notification just after the crash is triggered:

```
00:22:43.478372|zmqBridgeMamaTransportImpl_monitorEvent:socket:154e370 name:namingPub value:47 event:1 desc:CONNECTED endpoint:tcp://10.255.146.1:47023 (transport.c:1638)
00:22:43.479186|zmqBridgeMamaTransportImpl_dispatchNamingMsg:Received endpoint msg: type=C prog=xla host=cstransact2.dev.nyfix.net uuid=c6c17f2e-d044-4a5a-9297-d9b9df2fb06f pid=19132 topic=_NAMING pub=tcp://10.255.146.2:44757 (transport.c:541)
00:22:43.479536|zmqBridgeMamaTransportImpl_monitorEvent:socket:15455f0 name:dataSub value:48 event:1 desc:CONNECTED endpoint:tcp://10.255.146.2:44757 (transport.c:1638)
00:22:43.480088|zmqBridgeMamaTransportImpl_monitorEvent:socket:15455f0 name:dataSub value:0 event:4096 desc:HANDSHAKE_SUCCEEDED endpoint:tcp://10.255.146.2:44757 (transport.c:1638)
00:22:43.480381|zmqBridgeMamaTransportImpl_sendEndpointsMsg:Sent endpoint msg: prog=check_status host=cstransact4.dev.nyfix.net pid=23122 pub=tcp://10.255.146.4:56739 (transport.c:1470)
SEGV
Signal caught.  Abort called to create core and exit.
00:22:43.480411|zmqBridgeMamaTransportImpl_monitorEvent:socket:154e370 name:namingPub value:0 event:4096 desc:HANDSHAKE_SUCCEEDED endpoint:tcp://10.255.146.1:47023 (transport.c:1638)
ABRT
Signal caught.  Upon return from sig handler core & exit will be done.
```

It looks like the handshake is arriving right around the time of the send, and it may be triggering the crash.  If that's true, and if `NULL` is a legitimate value for `current`, maybe it would make sense to change the code as follows, to explicitly check for `NULL` and return:

```
void zmq::mtrie_t::match (unsigned char *data_, size_t size_,
    void (*func_) (pipe_t *pipe_, void *arg_), void *arg_)
{
    mtrie_t *current = this;
    while (current) {

        //  Signal the pipes attached to this node.
        if (current->pipes) {
            for (pipes_t::iterator it = current->pipes->begin ();
                  it != current->pipes->end (); ++it)
                func_ (*it, arg_);
        }

```

Again, I realize that it's asking a lot to diagnose this problem without a proper repro, but I'm hoping that someone with more knowledge of internals of ZeroMQ might be able to hazard a swag...

**Comment by sigiesec** [comment]
> Is there a legitimate way that current can be assigned a NULL value?

I don't think so. Something is inconsistent.

Maybe this is related to this problem reported by the clang static analyzer: https://travis-ci.org/zeromq/libzmq/jobs/392331396#L3752

Can you use libzmq 4.2.5? mtrie_t was changed into a template between 4.2.3 and 4.2.5

**Comment by WallStProg** [comment]
Woops!  "We have met the enemy and he is us".

Just got another crash and the stack trace on this one seems to point at my code being the cause.  Working to confirm and will post here when I'm able to. 

OTOH, still somewhat concerned with the warnings from clang-tidy, and would love to be able to write them off as false positives -- does anyone have any thoughts one way or the other?

Thanks!

**Comment by sigiesec** [comment]
@WallStProg Did you manage to confirm that the issue was caused by a bug in your client code?

The clang-tidy warnings should be checked thoroughly, yes, but this is a sort of different issue. And yes, the details are currently hidden due to the output format. We could generate HTML output instead, but it needs to be published somewhere. When I set up clang-tidy I was hoping that we could publish to SonarQube, but I did not find a cloud instance that supports the C++ community plugin, which is required to parse the clang-tidy output :( If you have any idea how to improve this, this would be very welcome.
